mesonbuild/options.py

```diff
@@ -682,96 +704,370 @@ def add_to_argparse(self, name: str, parser: argparse.ArgumentParser, help_suffi
+        #
+        # I did not do this yet, because it would make this MR even
+        # more massive than it already is. Later then.
+        if not self.is_cross and key.machine == MachineChoice.BUILD:
+            key = key.evolve(machine=MachineChoice.HOST)
         return key
 
-    def get_value_object(self, key: T.Union[OptionKey, str]) -> 'UserOption[T.Any]':
-        return self.d[self.ensure_key(key)]
-
     def get_value(self, key: T.Union[OptionKey, str]) -> 'T.Any':
         return self.get_value_object(key).value
 
+    def get_value_object_for(self, key: 'T.Union[OptionKey, str]') -> 'UserOption[T.Any]':
+        key = self.ensure_and_validate_key(key)
+        potential = self.options.get(key, None)
+        if self.is_project_option(key):
+            assert key.subproject is not None
+            if potential is not None and potential.yielding:
+                parent_key = key.evolve(subproject='')
+                parent_option = self.options[parent_key]
+                # If parent object has different type, do not yield.
+                # This should probably be an error.
+                if type(parent_option) is type(potential):
+                    return parent_option
+                return potential
+            if potential is None:
+                raise KeyError(f'Tried to access nonexistant project option {key}.')
+            return potential
+        else:
+            if potential is None:
+                parent_key = key.evolve(subproject=None)
+                if parent_key not in self.options:
+                    raise KeyError(f'Tried to access nonexistant project parent option {parent_key}.')
+                return self.options[parent_key]
+            return potential
+
+    def get_value_object(self, key: T.Union[OptionKey, str]) -> 'UserOption[T.Any]':
+        key = self.ensure_and_validate_key(key)
+        return self.options[key]
+
+    def get_value_object_and_value_for(self, key: OptionKey):
+        assert isinstance(key, OptionKey)
+        vobject = self.get_value_object_for(key)
+        computed_value = vobject.value
+        if key.subproject is not None:
+            keystr = str(key)
+            if keystr in self.augments:
+                computed_value = vobject.validate_value(self.augments[keystr])
+        return (vobject, computed_value)
+
+    def get_value_for(self, name: 'T.Union[OptionKey, str]', subproject: T.Optional[str] = None) -> 'OptionValueType':
+        if isinstance(name, str):
+            key = OptionKey(name, subproject)
+        else:
+            assert subproject is None
+            key = name
+        vobject, resolved_value = self.get_value_object_and_value_for(key)
+        return resolved_value
+
+    def num_options(self):
+        basic = len(self.options)
+        build = len(self.build_options) if self.build_options else 0
+        return basic + build
+
+    def __len__(self) -> int:
+        return len(self.d)
```

Comment:
```suggestion
        return len(self.options)
```
Looks like the rest of these got renamed?